share/source.F90

```diff
@@ -102,19 +104,22 @@ end function namchg
           ! this is a type 2 problem, skip
           ! reset the icl_tmp to -1 because of the type 2 not found
           icl_tmp = -1
+          count_2 = count_2 + 1
+          return
```

Comment:
The exit point of this method moved inside the conditional... doesn’t that change the behavior?

Reply:
Well, I tested with it. It turns out that `return` is the right one. The previous `exit` only exit the current block and goes to the following code, which slows down the `find_cell`.

Reply:
But there is no way to leave the `mxa` loop if a match is found.  It looks like it will always test every volume in the problem!

Reply:
Maybe you need both exit points?  One to exit with a type 2 error and one to exit with a successful match?

Reply:
Actually, I think the previous exit point should be faster.  If you find a matching cell with `chkcel` you always want to immediately return, either with that cell ID or with the error.

Reply:
> Actually, I think the previous exit point should be faster. If you find a matching cell with `chkcel` you always want to immediately return, either with that cell ID or with the error.

Yes, this is what we want. But the `exit` statement just ends the current `do` loop. The following type 3 check will still be running and is always true if a type 2 warning is found. this causes the wrong count of type3 error.

The `return` here is used to end of the entire `find_cell` subroutine and returns the current output value. The `return` statement seems to be obsolete in Fortran version newer than 95. 

There needs another variable to identify the warning type to differentiate type 2 and 3 if the `exit` is used.

Reply:
What about a `return` statement where the `exit` statement used to be?

I don't think `return` is obsolete.  I think it is obsolete to provide a return value in a `return` statement.

Reply:
> What about a `return` statement where the `exit` statement used to be?
> 
> I don't think `return` is obsolete. I think it is obsolete to provide a return value in a `return` statement.

Well, in  this PR, it is a `return` statement to replace the `exit`.
And for the `return` statement, I checked the tutorial I am referring: http://wwweic.eri.u-tokyo.ac.jp/computer/manual/altix/compile/Fortran/Intel_Fdoc100/main_for/mergedProjects/lref_for/source_files/rfreturn.htm You are right, it says `alternate return is an obsolescent feature in Fortran 95 and Fortran 90`. It's `altrnate return` (with a return value) not `return`.

Reply:
I don't have a problem with `return`, but it's in the wrong place.  It should be outside the `if` block

Reply:
Updated in https://github.com/pyne/pyne/pull/1159/commits/646b00c01be1073dacc3fb93a045d34cd7a82768.